docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,282 @@
+
+```
+GRANT BINLOG_ADMIN TO u1; // grants the role binlog_admin
+GRANT BINLOG_ADMIN ON *.* TO u1; // grants the dynamic privilege binlog_admin
+```
+
+This same nuance applies to MySQL.
+
+### Documentation Plan
+
+The statement reference pages for each of the affected metadata commands will need to be updated to describe dynamic privileges.
+
+There will also need to be documentation specific to `DYNAMIC` privileges to describe how it works, and the purpose of fine-grained access control.
+
+
+## Test Design
+
+Testing dynamic privileges is a little bit complex because of the various ways privileges can be inherited by users:
+
+* Direct `GRANT` to the user
+* Granting to a role that the user inherits.
+
+### Functional Tests
+
+Unit tests will be added to cover the semantics around role/dynamic privilege precedence, including logical restoring in a different order.
+
+Integration testing needs to test with global kill enabled/disabled.
+
+### Scenario Tests
+
+The use-cases required by the DBaaS team should be validated when combined with `security-enhanced-mode`. They are:
+
+| Account Name | root | cloudAdmin |
+| --------------- | --------------- | --------------- |
+| Backup & Restore to cloud | Y | Y |
+| File privilege | N | N |
+| Read or set variables | Y | Y | 
+| set restricted variables(some of them even can not be read) | N | Y |
+| Read or set restricted system tables | N | Y | 
+| DROP USER cloudAdmin | N | Hard to N(unless some hardcoded) |
+| REVOKE cloudAdmin | N | Hard to N(unless some hardcoded) |
+| Show processlist, Access to threads belong to other users | Y | Y |
+| Change password if the password of SUPER user is forgotten | N | Y |
+| Kill connections belong to cloudAdmin | N | Y |
+| SHUTDOWN / RESTART | N | Y (graceful shutdown on k8s for tidb-server) |
+
+### Compatibility Tests
+
+The introduction of `DYNAMIC` privileges is not expected to introduce any compatibility issues, because backwards compatibility is ensured. However, plugins should migrate to registering their own dynamic privileges and not rely on the use of `SUPER`. This is considered an enhancement, and not included in-scope for the initial introduction of dynamic privileges (which introduces the framework for plugins to use).
+
+## Impact & Risks
+
+In its initial release, dynamic privilege usage will be controlled by an experimental feature flag. The implementation will be via restricting `GRANT` and `REVOKE` statements from creating dynamic privileges (it is too invasive to conditionally modify the ast visitor functionaliy).
+
+For backwards compatibility, the MySQL-compatible dynamic privileges will also permit `SUPER`. This helps prevent upgrade issues, such a when TiDB was bootstrapped `GRANT ALL ON *.*` would not have granted all the dynamic privileges. There might be some impact on Upgrade/Downgrade story if eventually the `BACKUP_ADMIN` privilege is used instead of `SUPER`, but for the initial release I am planning to allow either.
```

Comment:
I'm curious whether there are requirements for MySQL compatible dynamic privileges at present, @IANTHEREAL do you have any idea?

Reply:
To paraphrase your question: is there a requirement where a user with `SUPER` but not a `DYNAMIC` privilege will be denied access to perform a task such as `BACKUP`. The main difference is:

* In MySQL, you can not `GRANT` a dynamic privilege `foo` unless you have the dynamic privilege `foo` with `grant_option`.
* In TiDB you need either the dynamic privilege `foo` with `grant option` or `super with grant option`.

This caters for the case that the `root` user already has `super`, but not each of the `dynamic privileges`. There is an existing incompatibility in TiDB where `grant all` does not actually truly grant all. I don't plan to fix it.

(Note: Slightly contradicting this, if SEM **is enabled**, there will be cases where `super` is locked down and a tidb-specific dynamic privilege is required. This is the main goal of SEM: restrict SUPER.)

Reply:
@zz-jason 
Yes, it has. On cloud, SRE should have the ability to perform BACKUP, without a SUPER privilege. So the BACKUP_ADMIN privilege should be granted to SRE’s account.

Reply:
Ah okay. Yes, consider the case that the backup is encrypted. You might have permission to configure the backup and restore, but no visibility into the data. This is a good feature to have.